### PR TITLE
net, rpc, cli: expose peer network in getpeerinfo; simplify/improve -netinfo

### PR DESCRIPTION
**Pull request by jonatack**
This PR:

- builds on #19991 and #19998
- exposes peer networks via a new getpeerinfo `network` field ("ipv4", "ipv6", or "onion"), and adds functional tests
- updates -netinfo to use getpeerinfo `network` rather than detecting the peer networks client-side
- refactors -netinfo to easily add future networks

### CODE REVIEW

**Comment by practicalswift** [comment]
Concept ACK

Thanks for improving `-netinfo`! :)

**Comment by jonatack** [comment]
This PR now builds on #19991 / #19998 and should be ready for review.

**Comment by hebasto** [comment]
Concept ACK.

**Review by hebasto — COMMENTED** [review]
Tested 5604a61435c5dfdeed41db5eca489b353dee0b99 on Linux Mint 20 (x86_64):

```
$ src/bitcoin-cli -netinfo 4
Bitcoin Core v0.20.99.0-5604a6143 - 70016/Satoshi:0.20.99/

Peer connections sorted by direction and min ping
<-> relay   net mping   ping send recv  txn  blk uptime  id address                     version
 in  full onion   178   1081   11   76              229 171 127.0.0.1:37328             70015/bitnodes.io:0.1/
 in block onion   241    558   29   29              265 157 127.0.0.1:34910             70015/Satoshi:0.20.1/
 in block onion   245    399  110  110              289 144 127.0.0.1:33322             70015/Satoshi:0.20.1/
 in  full onion   359    430   10    1    1   37     44 268 127.0.0.1:49482             70015/Satoshi:0.19.1/
out  full onion   251    450    1    3    0         468  78 jqsx6ag4hujpapxq.onion:8333 70015/Satoshi:0.20.0/
out  full onion   261    352    1    1    0   12    625   6 tuwj7ju4s25345pg.onion:8333 70015/Satoshi:0.19.0.1/
out block onion   280    528   58   58       624    625   9 kzrnyo7fadxihvgb.onion:8333 70015/Satoshi:0.20.0/
out  full onion   329    474    2    5    0  623    625   7 2mmxouhv6nebowkq.onion:8333 70015/Satoshi:0.18.1/
out  full onion   372    622    3    3    1  623    625   4 q467hmvzy5kfvoed.onion:8333 70015/Satoshi:0.20.1/
out  full onion   380    546    2   11    0  624    625   8 ee5skvb3ktbposl3.onion:8333 70015/Satoshi:0.20.0/
out  full onion   392    623    0    6    0  624    626   2 6xiqqr4fxnstd5fo.onion:8333 70015/Satoshi:0.20.1/
out  full onion   421    556    2    7    0  267    625   5 ohrieqovuxe6y4gl.onion:8333 70015/Satoshi:0.20.0/
out block onion   465    578  112  111              246 162 a53vtdm7uiet5vdl.onion:8333 70015/Satoshi:0.19.1/
out  full onion  1717   1717    5    0    0           1 290 mbciulcc3wuksb3f.onion:8333 70015/Satoshi:0.20.0/
                   ms     ms  sec  sec  min  min    min

        ipv4    ipv6   onion   total  block-relay
in         0       0       4       4       2
out        0       0      10      10       2
total      0       0      14      14       4
...
```

File: src/rpc/net.cpp [hunk 1]
```diff
@@ -94,6 +94,7 @@ static RPCHelpMan getpeerinfo()
                             {RPCResult::Type::STR, "addr", "(host:port) The IP address and port of the peer"},
                             {RPCResult::Type::STR, "addrbind", "(ip:port) Bind address of the connection to the peer"},
                             {RPCResult::Type::STR, "addrlocal", "(ip:port) Local address as reported by the peer"},
+                            {RPCResult::Type::STR, "network", "Network (ipv4, ipv6, or onion) the peer connected through"},
```
Comment (hebasto) [inline src/rpc/net.cpp:97]: bebd062eef3fa2d33c9216c368977dbedb70a443

While here, mind making "addrbind" and "addrlocal" ordering consistent for help and output?

File: test/functional/feature_proxy.py [hunk 2]
```diff
@@ -143,6 +162,7 @@ def node_test(self, node, proxies, auth, test_onion=True):
             assert_equal(cmd.username, None)
             assert_equal(cmd.password, None)
         rv.append(cmd)
+        self.network_test(node, addr, network=NET_UNROUTABLE)
```
Comment (hebasto) [inline test/functional/feature_proxy.py:165]: bebd062eef3fa2d33c9216c368977dbedb70a443

Testing of `network=NET_UNROUTABLE` is actually skipped. Could we avoid such skipping?

File: src/bitcoin-cli.cpp [hunk 3]
```diff
@@ -298,30 +296,11 @@ class GetinfoRequestHandler: public BaseRequestHandler
...
-    }
-    bool IsOutboundOnion(const std::string& addr, int mapped_as) const
-    {
-        const size_t addr_len{addr.size()};
-        const size_t onion_pos{addr.rfind(ONION)};
-        return mapped_as == 0 && onion_pos != std::string::npos && addr_len > ONION_LEN &&
-               (onion_pos == addr_len - ONION_LEN || onion_pos == addr.find_last_of(":") - ONION_LEN);
-    }
-    uint8_t m_details_level{0}; //!< Optional user-supplied arg to set dashboard details level
+    static const int m_networks = 3; // Number of possible networks
```
Comment (hebasto) [inline src/bitcoin-cli.cpp:299]: 5604a61435c5dfdeed41db5eca489b353dee0b99
```suggestion
    static constexpr int m_networks = 3; //!< Number of possible networks.
```

File: src/bitcoin-cli.cpp [hunk 4]
```diff
@@ -298,30 +296,11 @@ class GetinfoRequestHandler: public BaseRequestHandler
...
+    uint8_t m_details_level = 0;     //!< Optional user-supplied arg to set dashboard details level
```
Comment (hebasto) [inline src/bitcoin-cli.cpp:300]: Why switch from modern `{}` initialization to the old style?

**Review by hebasto — APPROVED** [review]
ACK 5604a61435c5dfdeed41db5eca489b353dee0b99

File: src/rpc/net.cpp (hunk 1 above)
Comment (jonatack) [reply src/rpc/net.cpp:97]: Good idea. Done.

File: src/bitcoin-cli.cpp (hunk 3 above)
Comment (jonatack) [reply src/bitcoin-cli.cpp]: Done

File: src/bitcoin-cli.cpp (hunk 4 above)
Comment (jonatack) [reply src/bitcoin-cli.cpp]: Done, retained modern style. Thanks.

File: test/functional/feature_proxy.py (hunk 2 above)
Comment (jonatack) [reply test/functional/feature_proxy.py:165]: When I run `feature_proxy.py` with `-l debug`, or with the following patch, it doesn't seem to be skipped in my testing. LMK if I'm misunderstanding.
```diff
+++ b/test/functional/feature_proxy.py
@@ -101,6 +101,7 @@ class ProxyTest(BitcoinTestFramework):
         for peer in node.getpeerinfo():
             if peer["addr"] == addr:
                 assert_equal(peer["network"], network)
+                import pprint; pprint.pprint(network)
```

**Comment by jonatack** [comment]
Thanks for reviewing @hebasto. I've addressed your feedback.

**Comment by laanwj** [comment]
Needs rebase after #19998.

**Comment by jonatack** [comment]
Rebased. If merged for 0.21 will add a release note manually in the wiki.

**Comment by jonatack** [comment]
Improved the last commit so upcoming additional networks can be added to -netinfo by updating 3 lines.

File: src/bitcoin-cli.cpp [hunk 5]
```diff
@@ -298,30 +296,22 @@ class GetinfoRequestHandler: public BaseRequestHandler
...
-    bool IsOutboundOnion(const std::string& addr, int mapped_as) const
-    {
-        const size_t addr_len{addr.size()};
-        const size_t onion_pos{addr.rfind(ONION)};
-        return mapped_as == 0 && onion_pos != std::string::npos && addr_len > ONION_LEN &&
-               (onion_pos == addr_len - ONION_LEN || onion_pos == addr.find_last_of(":") - ONION_LEN);
+        for (const auto& network : m_networks) {
+            if (str == network.first) return network.second;
+        }
+        assert(false);
```
Comment (laanwj) [inline src/bitcoin-cli.cpp:306]: Let's not do `assert(false)` here. A crash on "invalid" import from the RPC is never good, it is good to be slightly forgiving with regard to forward-compatibility. Maybe return an 'unknown' category instead that is not used for the statistics/table.

File: src/interfaces/node.h [hunk 6]
```diff
@@ -23,7 +23,7 @@
 class BanMan;
 class CCoinControl;
 class CFeeRate;
-class CNodeStats;
+struct CNodeStats;
```
Comment (laanwj) [inline src/interfaces/node.h:26]: ~0 on the change from a class to a struct, I think calling it a struct is more consistent with how it's used but it doesn't seem to be a logical part of this change

**Comment by laanwj** [comment]
> Rebased. If merged for 0.21 will add a release note manually in the wiki.

+1 in getting this into 0.21, it's a low risk change and is kind of the culmination of work that's been going on for the entire release cycle

**Review by hebasto — COMMENTED** [review]
Tested 6c08a4b6e755efa40f021ff4e7ec0d69379cc773 on Linux Mint 20 (x86_64). LGTM.

File: src/bitcoin-cli.cpp [hunk 7]
```diff
@@ -298,30 +296,22 @@ class GetinfoRequestHandler: public BaseRequestHandler
 class NetinfoRequestHandler : public BaseRequestHandler
 {
 private:
-    bool IsAddrIPv6(const std::string& addr) const
+    const std::vector<std::pair<std::string, uint8_t>> m_networks{{"ipv4", 0}, {"ipv6", 1}, {"onion", 2}};
```
Comment (laanwj) [inline src/bitcoin-cli.cpp:299]: As these IDs are always consecutively, and you iterate over it linearly, why not simply
```
const std::vector<std::string> m_networks{"ipv4","ipv6","onion"};
```
alternatively, *lookup* would be easier with a `std::map`.

File: src/bitcoin-cli.cpp (hunk 5 above)
Comment (jonatack) [reply src/bitcoin-cli.cpp]: Done!

File: src/interfaces/node.h (hunk 6 above)
Comment (jonatack) [reply src/interfaces/node.h]: Removed the struct change.

File: src/bitcoin-cli.cpp (hunk 7 above)
Comment (jonatack) [reply src/bitcoin-cli.cpp]: Good idea--done, and changed the new data structures in the last commit to fixed-sized `std::array`.

Comment (jonatack) [reply src/bitcoin-cli.cpp]: Good idea to flatten it--done, and changed the new data structures in the last commit to fixed-sized std::arrays.

**Comment by jonatack** [comment]
Thanks for the feedback!

<details><summary>Updated per <code>git diff 6c08a4b b62d8a4</code></summary><p>

```diff
diff --git a/src/bitcoin-cli.cpp b/src/bitcoin-cli.cpp
index e44aceddfc..f73b68a52a 100644
--- a/src/bitcoin-cli.cpp
+++ b/src/bitcoin-cli.cpp
@@ -296,14 +296,15 @@ public:
 class NetinfoRequestHandler : public BaseRequestHandler
 {
 private:
-    const std::vector<std::pair<std::string, uint8_t>> m_networks{{"ipv4", 0}, {"ipv6", 1}, {"onion", 2}};
-    const size_t m_networks_size{m_networks.size()};
-    uint8_t NetworkStringToId(const std::string& str) const
+    static constexpr int8_t UNKNOWN_NETWORK{-1};
+    static constexpr size_t m_networks_size{3};
+    const std::array<std::string, m_networks_size> m_networks{"ipv4", "ipv6", "onion"};
+    int8_t NetworkStringToId(const std::string& str) const
     {
-        for (const auto& network : m_networks) {
-            if (str == network.first) return network.second;
+        for (size_t i = 0; i < m_networks_size; ++i) {
+            if (str == m_networks.at(i)) return i;
         }
-        assert(false);
+        return UNKNOWN_NETWORK;
     }
     uint8_t m_details_level{0}; //!< Optional user-supplied arg to set dashboard details level
     bool DetailsRequested() const { return m_details_level > 0 && m_details_level < 5; }
@@ -369,10 +370,11 @@ public:
         }
 
         // Count peer connection totals, and if DetailsRequested(), store peer data in a vector of structs.
-        std::vector<std::vector<int>> counts(3, std::vector<int>(m_networks_size + 2, 0)); // conn counts by (in/out/total, ipv4/ipv6/onion/total/block-relay)
+        std::array<std::array<uint16_t, m_networks_size + 2>, 3> counts{}; // conn counts by (in/out/total, networks/total/block-relay)
         for (const UniValue& peer : batch[ID_PEERINFO]["result"].getValues()) {
             const std::string network{peer["network"].get_str()};
-            const uint8_t network_id{NetworkStringToId(network)};
+            const int8_t network_id{NetworkStringToId(network)};
+            if (network_id == UNKNOWN_NETWORK) continue;
             const bool is_outbound{!peer["inbound"].get_bool()};
             const bool is_block_relay{!peer["relaytxes"].get_bool()};
             ++counts.at(is_outbound).at(network_id);      // in/out by network
@@ -440,7 +442,7 @@ public:
 
         // Report peer connection totals by type.
         result += "        ipv4    ipv6   onion   total  block-relay\n";
-        const std::vector<std::string> rows{"in", "out", "total"};
+        const std::array<std::string, 3> rows{"in", "out", "total"};
         for (size_t i = 0; i < m_networks_size; ++i) {
             result += strprintf("%-5s  %5i   %5i   %5i   %5i   %5i\n", rows.at(i), counts.at(i).at(0), counts.at(i).at(1), counts.at(i).at(2), counts.at(i).at(m_networks_size), counts.at(i).at(m_networks_size + 1));
         }
diff --git a/src/interfaces/node.h b/src/interfaces/node.h
index 27d8740812..5079be038e 100644
--- a/src/interfaces/node.h
+++ b/src/interfaces/node.h
@@ -23,7 +23,7 @@
 class BanMan;
 class CCoinControl;
 class CFeeRate;
-struct CNodeStats;
+class CNodeStats;
 class Coin;
 class RPCTimerInterface;
 class UniValue;
diff --git a/src/net.h b/src/net.h
index 2435353261..eec48a7e95 100644
--- a/src/net.h
+++ b/src/net.h
@@ -100,7 +100,7 @@ struct AddedNodeInfo
     bool fInbound;
 };
 
-struct CNodeStats;
+class CNodeStats;
 class CClientUIInterface;
 
 struct CSerializedNetMsg
@@ -672,8 +672,9 @@ extern std::map<CNetAddr, LocalServiceInfo> mapLocalHost GUARDED_BY(cs_mapLocalH
 extern const std::string NET_MESSAGE_COMMAND_OTHER;
 typedef std::map<std::string, uint64_t> mapMsgCmdSize; //command, total bytes
 
-struct CNodeStats
+class CNodeStats
 {
+public:
     NodeId nodeid;
     ServiceFlags nServices;
     bool fRelayTxes;
@@ -705,8 +706,9 @@ struct CNodeStats
     CAddress addr;
     // Bind address of our side of the connection
     CAddress addrBind;
-    uint32_t m_mapped_as;
+    // Name of the network the peer connected through
     std::string m_network;
+    uint32_t m_mapped_as;
     std::string m_conn_type_string;
 }; 
```
</p></details>

**Comment by laanwj** [comment]
ACK 6272604bef3b409455b010d134b4b62c8f6ff49f

**Comment by jonatack** [comment]
> Rebased. If merged for 0.21 will add a release note manually in the wiki.

Done, manually added a release note in the wiki for getpeerinfo#network.

**Pull request merged**